doc/man-pages/man1/udebug.1
.\" Automatically generated by Pod::Man 2.16 (Pod::Simple 3.05)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "UDEBUG 1"
.TH UDEBUG 1 "2010-02-11" "OpenAFS" "AFS Command Reference"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
udebug \- Reports Ubik process status for a database server process
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBudebug\fR \fB\-servers\fR\ <\fIserver\ machine\fR> [\fB\-port\fR\ <\fI\s-1IP\s0\ port\fR>]
    [\fB\-long\fR] [\fB\-help\fR]
.PP
\&\fBudebug\fR \fB\-s\fR\ <\fIserver\ machine\fR> [\fB\-p\fR\ <\fI\s-1IP\s0\ port\fR>] [\fB\-l\fR] [\fB\-h\fR]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The \fBudebug\fR command displays the status of the lightweight Ubik process
for the database server process identified by the \fB\-port\fR argument that
is running on the database server machine named by the \fB\-servers\fR
argument. The output identifies the machines where peer database server
processes are running, which of them is the synchronization site (Ubik
coordinator), and the status of the connections between them.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-servers\fR <\fIserver machine\fR>" 4
.IX Item "-servers <server machine>"
Names the database server machine that is running the process for which to
display status information. Provide the machine's \s-1IP\s0 address in dotted
decimal format, its fully qualified host name (for example,
\&\fBfs1.abc.com\fR), or the shortest abbreviated form of its host name that
distinguishes it from other machines. Successful use of an abbreviated
form depends on the availability of a name resolution service (such as the
Domain Name Service or a local host table) at the time the command is
issued.
.IP "\fB\-port\fR <\fI\s-1IP\s0 port\fR>" 4
.IX Item "-port <IP port>"
Identifies the database server process for which to display status
information, either by its process name or port number. Provide one of the
following values.
.RS 4
.IP "\fBbuserver\fR or 7021 for the Backup Server" 4
.IX Item "buserver or 7021 for the Backup Server"
.PD 0
.IP "\fBkaserver\fR or 7004 for the Authentication Server" 4
.IX Item "kaserver or 7004 for the Authentication Server"
.IP "\fBptserver\fR or 7002 for the Protection Server" 4
.IX Item "ptserver or 7002 for the Protection Server"
.IP "\fBvlserver\fR or 7003 for the Volume Location Server" 4
.IX Item "vlserver or 7003 for the Volume Location Server"
.RE
.RS 4
.RE
.IP "\fB\-long\fR" 4
.IX Item "-long"
.PD
Reports additional information about each peer of the machine named by the
\&\fB\-servers\fR argument. The information appears by default if that machine
is the synchronization site.
.IP "\fB\-help\fR" 4
.IX Item "-help"
Prints the online help for this command. All other valid options are
ignored.
.SH "OUTPUT"
.IX Header "OUTPUT"
Several of the messages in the output provide basic status information
about the Ubik process on the machine specified by the \fB\-servers\fR
argument, and the remaining messages are useful mostly for debugging
purposes.
.PP
To check basic Ubik status, issue the command for each database server
machine in turn. In the output for each, one of the following messages
appears in the top third of the output.
.PP
.Vb 1
\&   I am sync site . . . (<#_sites> servers)
\&
\&   I am not sync site
.Ve
.PP
For the synchronization site, the following message indicates that all
sites have the same version of the database, which implies that Ubik is
functioning correctly. See the following for a description of values other
than \f(CW\*(C`1f\*(C'\fR.
.PP
.Vb 1
\&   Recovery state 1f
.Ve
.PP
For correct Ubik operation, the database server machine clocks must agree
on the time. The following messages, which are the second and third lines
in the output, report the current date and time according to the database
server machine's clock and the clock on the machine where the \fBudebug\fR
command is issued.
.PP
.Vb 2
\&   Host\*(Aqs <IP_addr> time is <dbserver_date/time>
\&   Local time is <local_date/time> (time differential <skew> secs)
.Ve
.PP
The <skew> is the difference between the database server machine clock and
the local clock. Its absolute value is not vital for Ubik functioning, but
a difference of more than a few seconds between the \fIskew\fR values for the
database server machines indicates that their clocks are not synchronized
and Ubik performance is possibly hampered.
.PP
Following is a description of all messages in the output. As noted, it is
useful mostly for debugging and most meaningful to someone who understands
Ubik's implementation.
.PP
The output begins with the following messages. The first message reports
the \s-1IP\s0 addresses that are configured with the operating system on the
machine specified by the \fB\-servers\fR argument. As previously noted, the
second and third messages report the current date and time according to
the clocks on the database server machine and the machine where the
\&\fBudebug\fR command is issued, respectively. All subsequent timestamps in
the output are expressed in terms of the local clock rather than the
database server machine clock.
.PP
.Vb 3
\&   Host\*(Aqs addresses are: <list_of_IP_addrs>
\&   Host\*(Aqs <IP_addr> time is <dbserver_date/time>
\&   Local time is <local_date/time> (time differential <skew> secs)
.Ve
.PP
If the <skew> is more than about 10 seconds, the following message
appears. As noted, it does not necessarily indicate Ubik malfunction: it
denotes clock skew between the database server machine and the local
machine, rather than among the database server machines.
.PP
.Vb 1
\&   ****clock may be bad
.Ve
.PP
If the udebug command is issued during the coordinator election process
and voting has not yet begun, the following message appears next.
.PP
.Vb 1
\&   Last yes vote not cast yet
.Ve
.PP
Otherwise, the output continues with the following messages.
.PP
.Vb 3
\&   Last yes vote for <sync_IP_addr> was <last_vote> secs ago (sync site);
\&   Last vote started <vote_start> secs ago (at <date/time>)
\&   Local db version is <db_version>
.Ve
.PP
The first indicates which peer this Ubik process last voted for as
coordinator (it can vote for itself) and how long ago it sent the vote.
The second message indicates how long ago the Ubik coordinator requested
confirming votes from the secondary sites. Usually, the <last_vote> and
<vote_start> values are the same; a difference between them can indicate
clock skew or a slow network connection between the two database server
machines. A small difference is not harmful. The third message reports the
current version number <db_version> of the database maintained by this
Ubik process. It has two fields separated by a period. The field before
the period is based on a timestamp that reflects when the database first
changed after the most recent coordinator election, and the field after
the period indicates the number of changes since the election.
.PP
The output continues with messages that differ depending on whether the
Ubik process is the coordinator or not.
.IP "\(bu" 4
If there is only one database server machine, it is always the coordinator
(synchronization site), as indicated by the following message.
.Sp
.Vb 1
\&   I am sync site forever (1 server)
.Ve
.IP "\(bu" 4
If there are multiple database sites, and the \fB\-servers\fR argument names
the coordinator (synchronization site), the output continues with the
following two messages.
.Sp
.Vb 3
\&   I am sync site until <expiration> secs from now (at <date/time>)
\&       (<#_sites> servers)
\&   Recovery state <flags>
.Ve
.Sp
The first message (which is reported on one line) reports how much longer
the site remains coordinator even if the next attempt to maintain quorum
fails, and how many sites are participating in the quorum. The \fIflags\fR
field in the second message is a hexadecimal number that indicates the
current state of the quorum. A value of \f(CW\*(C`1f\*(C'\fR indicates complete database
synchronization, whereas a value of \f(CW\*(C`f\*(C'\fR means that the coordinator has
the correct database but cannot contact all secondary sites to determine
if they also have it. Lesser values are acceptable if the \fBudebug\fR
command is issued during coordinator election, but they denote a problem
if they persist. The individual flags have the following meanings:
.RS 4
.IP "0x1" 4
.IX Item "0x1"
This machine is the coordinator.
.IP "0x2" 4
.IX Item "0x2"
The coordinator has determined which site has the database with the
highest version number.
.IP "0x4" 4
.IX Item "0x4"
The coordinator has a copy of the database with the highest version
number.
.IP "0x8" 4
.IX Item "0x8"
The database's version number has been updated correctly.
.IP "0x10" 4
.IX Item "0x10"
All sites have the database with the highest version number.
.RE
.RS 4
.Sp
If the udebug command is issued while the coordinator is writing a change
into the database, the following additional message appears.
.Sp
.Vb 1
\&   I am currently managing write transaction I<identifier>
.Ve
.RE
.IP "\(bu" 4
If the \fB\-servers\fR argument names a secondary site, the output continues
with the following messages.
.Sp
.Vb 3
\&   I am not sync site
\&   Lowest host <lowest_IP_addr> was set <low_time> secs ago
\&   Sync host <sync_IP_addr> was set <sync_time> secs ago
.Ve
.Sp
The <lowest_IP_addr> is the lowest \s-1IP\s0 address of any peer from which the
Ubik process has received a message recently, whereas the <sync_IP_addr>
is the \s-1IP\s0 address of the current coordinator. If they differ, the machine
with the lowest \s-1IP\s0 address is not currently the coordinator. The Ubik
process continues voting for the current coordinator as long as they
remain in contact, which provides for maximum stability. However, in the
event of another coordinator election, this Ubik process votes for the
<lowest_IP_addr> site instead (assuming they are in contact), because it
has a bias to vote in elections for the site with the lowest \s-1IP\s0 address.
.PP
For both the synchronization and secondary sites, the output continues
with the following messages. The first message reports the version number
of the database at the synchronization site, which needs to match the
<db_version> reported by the preceding \f(CW\*(C`Local db version\*(C'\fR message. The
second message indicates how many \s-1VLDB\s0 records are currently locked for
any operation or for writing in particular. The values are nonzero if the
\&\fBudebug\fR command is issued while an operation is in progress.
.PP
.Vb 2
\&   Sync site\*(Aqs db version is <db_version>
\&   <locked> locked pages, <writes> of them for write
.Ve
.PP
The following messages appear next only if there are any read or write
locks on database records:
.PP
.Vb 2
\&   There are read locks held
\&   There are write locks held
.Ve
.PP
Similarly, one or more of the following messages appear next only if there
are any read or write transactions in progress when the \fBudebug\fR command
is issued:
.PP
.Vb 3
\&   There is an active write transaction
\&   There is at least one active read transaction
\&   Transaction tid is <tid>
.Ve
.PP
If the machine named by the \fB\-servers\fR argument is the coordinator, the
next message reports when the current coordinator last updated the
database.
.PP
.Vb 2
\&    Last time a new db version was labelled was:
\&            <last_restart> secs ago (at <date/time>)
.Ve
.PP
If the machine named by the \fB\-servers\fR argument is the coordinator, the
output concludes with an entry for each secondary site that is
participating in the quorum, in the following format.
.PP
.Vb 5
\&   Server (<IP_address>): (db <db_version>)
\&   last vote rcvd <last_vote> secs ago (at <date/time>),
\&   last beacon sent <last_beacon> secs ago (at <date/time>),
\&       last vote was { yes | no }
\&   dbcurrent={ 0 | 1 }, up={ 0 | 1 } beaconSince={ 0 | 1 }
.Ve
.PP
The first line reports the site's \s-1IP\s0 address and the version number of the
database it is maintaining. The <last_vote> field reports how long ago the
coordinator received a vote message from the Ubik process at the site, and
the <last_beacon> field how long ago the coordinator last requested a vote
message. If the \fBudebug\fR command is issued during the coordinator
election process and voting has not yet begun, the following messages
appear instead.
.PP
.Vb 2
\&   Last vote never rcvd
\&   Last beacon never sent
.Ve
.PP
On the final line of each entry, the fields have the following meaning:
.IP "\(bu" 4

NEWS
OpenAFS News -- history of user-visible changes.

* Changes incorporated in OpenAFS 1.4.0

** Addition of pthreading support in volserver.

** Threading optimization of RX libraries.

** Windows 95/98 client desupported.

** Digital UNIX/Tru64 desupported.

** Windows XP/2003 client includes persistent (across boots) cache support.

** Windows XP/2003 now supports a cache of up to 1.2gb.

** Windows XP/2003 has various improvements to the included GUI and command 
   line tools.

** Windows XP/2003 now supports AFS UNC paths.

** MacOS X (10.3) and Linux use system inodes instead of a private inode pool.

** Files larger than 2gb are supported in Unix clients, and in servers if the 
   --enable-largefile-fileserver switch is enabled at configure time.

** Obsolete and insecure auxiliary network daemons (rcp, rsh, ftpd et al)
   have been removed.

** Fileserver optimizations have been made to improve timeout and callback 
   breaking behavior to inaccessible clients or clients whose address has
   changed.

** Kerberos 5 support is integrated.

** -nosettime is now the default for afsd.  Use "-settime" to get the
   old behavior.

** Mountpoint directory information is now only faked for cross-cell
   mountpoints when using the -fakestat flag (e.g. for the directories
   under /afs, but not for most other volumes mounted inside the cell).
   The -fakestat-all switch can be used to fake information for all
   mountpoints.

** When fakestat is enabled on MacOSX, the Finder can be used to browse
   a fully-populated /afs directory.  However, this precludes reliable
   use of entire volumes as MacOS bundles (i.e. containing a Contents
   directory in the root of the volume).

** Mountpoint directory information can be faked by the cache manager,
   making operations such as stat'ing all cells under /afs much faster.
   This is enabled by passing -fakestat to afsd, but might not be stable
   on all platforms.

* Changes incorporated in OpenAFS 1.2.9

** The kaserver now defaults to not allowing interrealm authentication,
   due to security vulnerabilities in the krb4 protocol.  The new
   "-crossrealm" flag to the kaserver is provided to reenable interrealm
   authentication if desired.

** RedHat Linux 9.0 is now supported.

** Solaris 9 12/02 is now supported.  Solaris 7 and 8 x86 should now
   work again.

** On Linux machines using 2.2 series kernels, 2.2.19 or higher is now
   required.

** An OpenAFS 1.2.9 afsd will not work with kernel modules built from
   an earlier OpenAFS release.  In general, using a mismatched afsd and
   kernel modules set is unsupported; it is not recommended that you use
   such a configuration on a regular basis.

* Changes incorporated in OpenAFS 1.2.8

** Mountpoint directory information is now only faked for cross-cell
   mountpoits when using the -fakestat flag (e.g. for the directories
   under /afs, but not for most other volumes mounted inside the cell).
   The -fakestat-all switch can be used to fake information for all
   mountpoints.

** HPUX 11.0 is now supported.

** It is now possible for AFS to use Kerberos 5 directly, via rxkad 2b.
   See the OpenAFS 1.2.8 Release Notes for more information on using this
   capability.

** An NFS translator kernel module is now included and compiled by default
   for Solaris only.

* Changes incorporated in OpenAFS 1.2.7

** MacOS X 10.2 is now supported.  FreeBSD 4.3 and later support is included
   in this release, but is still under active development and should only
   be used by those doing active development on the OpenAFS FreeBSD client.

** When fakestat is enabled on MacOSX, the Finder can be used to browse
   a fully-populated /afs directory.  However, this precludes reliable
   use of entire volumes as MacOS bundles (i.e. containing a Contents
   directory in the root of the volume).

** The fileserver will now use Rx pings to determine if clients are reachable
   prior to allocating resources to them, to prevent asymmetric clients from
   consuming all fileserver resources.

* Changes incorporated in OpenAFS 1.2.6

** Mountpoint directory information can be faked by the cache manager,
   making operations such as stat'ing all cells under /afs much faster.
   This is enabled by passing -fakestat to afsd.

** Solaris 9 FCS and Solaris 7 and 8 x86 are now supported.

* Changes incorporated in OpenAFS 1.2.5

** A remote denial of service attack in the AIX and IRIX clients has
   been fixed.  Users of those platforms are strongly encouraged to
   upgrade.

** Fixed race conditions in fileserver that could result in crash.

* Changes incorporated in OpenAFS 1.2.4

** Server logfiles now more consistant about format in which hosts are
   referred to.

** vfsck on Solaris will now allow force runs (using -y flag) even if old 
   inodes exist.

* Changes incorporated in OpenAFS 1.2.3

** Cell aliases for dynroot can be specified in the CellAlias file in
   /usr/vice/etc or /usr/local/etc/openafs, in format "realname alias",
   one per line.  They can also be managed at runtime with "fs newalias"
   and "fs listaliases".

* Changes incorporated in OpenAFS 1.2.2

** Solaris 9 and Linux PA-RISC are now supported

** fileserver will not erroneously delay legitimate errors for 3 seconds
   after 10 errors are returned (e.g. stat() on a directory you can't read)

** Rx MTU calculation now works for Irix, Solaris and Linux

** If afsd is started with the -dynroot flag, /afs will be locally
   generated from the CellServDB.  AFSDB cells will be mounted
   automatically upon access.

** The namei fileserver allows vice "partitions" to be directories instead
   of partitions and will attach and display accordingly. Creating the file
   "AlwaysAttach" in the /vicepX directory is used as the trigger to attach it.

** TSM support for butc no longer requires editing a Makefile, simply
   specify the --enable-tivoli-tsm configure option.

** Linux builds no longer require source changes every time the kernel

src/ubik/udebug.c
/*
 * Copyright 2000, International Business Machines Corporation and others.
 * All Rights Reserved.
 * 
 * This software has been released under the terms of the IBM Public
 * License.  For details, see the LICENSE file in the top-level source
 * directory or online at http://www.openafs.org/dl/license10.html
 */

#include <afsconfig.h>
#include <afs/param.h>


#include <sys/types.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <signal.h>
#include <time.h>

#ifdef AFS_NT40_ENV
#include <winsock2.h>
#else
#include <sys/file.h>
#include <sys/param.h>
#include <netinet/in.h>
#include <netdb.h>
#endif

#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif

#include <lock.h>
#include <rx/xdr.h>
#include <rx/rx.h>
#include <afs/cmd.h>
#include <afs/afsutil.h>

#define UBIK_INTERNALS
#include "ubik.h"
#include "ubik_int.h"

static_inline int afs_cast_time_t(time_t d) { return (int) d; }

static short
PortNumber(register char *aport)
{
    register int tc;
    register afs_int32 total;

    total = 0;
    while ((tc = *aport++)) {
	if (tc < '0' || tc > '9')
	    return -1;		/* bad port number */
	total *= 10;
	total += tc - (int)'0';
    }
    return (total);
}

static short
PortName(char *aname)
{
    struct servent *ts;
    int len;

    ts = getservbyname(aname, NULL);

    if (ts)
	return ntohs(ts->s_port);	/* returns it in host byte order */

    len = strlen(aname);
    if (strncmp(aname, "vlserver", len) == 0) {
	return 7003;
    } else if (strncmp(aname, "ptserver", len) == 0) {
	return 7002;
    } else if (strncmp(aname, "kaserver", len) == 0) {
	return 7004;
    } else if (strncmp(aname, "buserver", len) == 0) {
	return 7021;
    }
    return (-1);
}

static int
CommandProc(struct cmd_syndesc *as, void *arock)
{
    char *hostName, *portName, *times;
    afs_int32 hostAddr;
    register afs_int32 i, j, code;
    short port;
    int int32p;
    time_t now, then, diff, newtime;
    struct hostent *th;
    struct rx_connection *tconn;
    struct rx_securityClass *sc;
    struct ubik_debug udebug;
    struct ubik_sdebug usdebug;
    int oldServer = 0;		/* are we talking to a pre 3.5 server? */
    afs_int32 isClone = 0;

    int32p = (as->parms[2].items ? 1 : 0);

    if (as->parms[0].items)
	hostName = as->parms[0].items->data;
    else
	hostName = NULL;

    if (as->parms[1].items)
	portName = as->parms[1].items->data;
    else
	portName = NULL;

    /* lookup host */
    if (hostName) {
	th = hostutil_GetHostByName(hostName);
	if (!th) {
	    printf("udebug: host %s not found in host table\n", hostName);
	    exit(1);
	}
	memcpy(&hostAddr, th->h_addr, sizeof(afs_int32));
    } else
	hostAddr = htonl(0x7f000001);	/* IP localhost */

    if (!portName)
	port = htons(3000);	/* default */
    else {
	port = PortNumber(portName);
	if (port < 0)
	    port = PortName(portName);
	if (port < 0) {
	    printf("udebug: can't resolve port name %s\n", portName);
	    exit(1);
	}
	port = htons(port);
    }

    rx_Init(0);
    sc = rxnull_NewClientSecurityObject();
    tconn = rx_NewConnection(hostAddr, port, VOTE_SERVICE_ID, sc, 0);

    /* now do the main call */
    code = VOTE_XDebug(tconn, &udebug, &isClone);
    if (code)
	code = VOTE_Debug(tconn, &udebug);
    if (code == RXGEN_OPCODE) {
	oldServer = 1;		/* talking to a pre 3.5 server */
	memset(&udebug, 0, sizeof(udebug));
	code = VOTE_DebugOld(tconn, (ubik_debug_old *)&udebug);
    }

    if (code) {
	printf("return code %d from VOTE_Debug\n", code);
	exit(0);
    }
    now = time(0);
    then = udebug.now;

    /* now print the main info */
    times = ctime(&then);
    times[24] = 0;
    if (!oldServer) {
	printf("Host's addresses are: ");
	for (j = 0; udebug.interfaceAddr[j] && (j < UBIK_MAX_INTERFACE_ADDR);
	     j++)
	    printf("%s ", afs_inet_ntoa(htonl(udebug.interfaceAddr[j])));
	printf("\n");
    }
    printf("Host's %s time is %s\n", afs_inet_ntoa(hostAddr), times);

    times = ctime(&now);
    times[24] = 0;
    diff = now - udebug.now;
    printf("Local time is %s (time differential %d secs)\n", times, afs_cast_time_t(diff));
    if (abs((int)diff) >= MAXSKEW)
	printf("****clock may be bad\n");

    /* UBIK skips the voting if 1 server - so we fudge it here */
    if (udebug.amSyncSite && (udebug.nServers == 1)) {
	udebug.lastYesHost = hostAddr;
	udebug.lastYesTime = udebug.now;
	udebug.lastYesState = 1;
	udebug.lastYesClaim = udebug.now;
	udebug.syncVersion.epoch = udebug.localVersion.epoch;
	udebug.syncVersion.counter = udebug.localVersion.counter;
    }

    /* XDR converts addresses for us, so all addresses are in HBO */
    if (udebug.lastYesHost == 0xffffffff) {
	printf("Last yes vote not cast yet \n");
    } else {
	diff = udebug.now - udebug.lastYesTime;
	printf("Last yes vote for %s was %d secs ago (%ssync site); \n",
	       afs_inet_ntoa(htonl(udebug.lastYesHost)),
	       afs_cast_time_t(diff),
	       ((udebug.lastYesState) ? "" : "not "));

	diff = udebug.now - udebug.lastYesClaim;
	newtime = now - diff;
	times = ctime(&newtime);
	times[24] = 0;
	printf("Last vote started %d secs ago (at %s)\n", 
	       afs_cast_time_t(diff), times);
    }

    printf("Local db version is %d.%d\n", udebug.localVersion.epoch,
	   udebug.localVersion.counter);

    if (udebug.amSyncSite) {
	if (udebug.syncSiteUntil == 0x7fffffff) {
	    printf("I am sync site forever (%d server%s)\n", udebug.nServers,
		   ((udebug.nServers > 1) ? "s" : ""));
	} else {
	    diff = udebug.syncSiteUntil - udebug.now;
	    newtime = now + diff;
	    times = ctime(&newtime);
	    times[24] = 0;
	    printf
		("I am sync site until %d secs from now (at %s) (%d server%s)\n",
		 afs_cast_time_t(diff), times, udebug.nServers,
		 ((udebug.nServers > 1) ? "s" : ""));
	}
	printf("Recovery state %x\n", udebug.recoveryState);
	if (udebug.activeWrite) {
	    printf("I am currently managing write trans %d.%d\n",
		   udebug.epochTime, udebug.tidCounter);
	}
    } else {
	if (isClone)
	    printf("I am a clone and never can become sync site\n");
	else
	    printf("I am not sync site\n");
	diff = udebug.now - udebug.lowestTime;
	printf("Lowest host %s was set %d secs ago\n",
	       afs_inet_ntoa(htonl(udebug.lowestHost)),
	       afs_cast_time_t(diff));

	diff = udebug.now - udebug.syncTime;
	printf("Sync host %s was set %d secs ago\n",
	       afs_inet_ntoa(htonl(udebug.syncHost)),
	       afs_cast_time_t(diff));
    }

    printf("Sync site's db version is %d.%d\n", udebug.syncVersion.epoch,
	   udebug.syncVersion.counter);
    printf("%d locked pages, %d of them for write\n", udebug.lockedPages,
	   udebug.writeLockedPages);

    if (udebug.anyReadLocks)
	printf("There are read locks held\n");
    if (udebug.anyWriteLocks)
	printf("There are write locks held\n");

    if (udebug.currentTrans) {
	if (udebug.writeTrans)
	    printf("There is an active write transaction\n");
	else
	    printf("There is at least one active read transaction\n");
	printf("Transaction tid is %d.%d\n", udebug.syncTid.epoch,
	       udebug.syncTid.counter);
    }
    if (udebug.epochTime) {
	diff = udebug.now - udebug.epochTime;
	newtime = now - diff;
	times = ctime(&newtime);
	times[24] = 0;
	printf
	    ("Last time a new db version was labelled was:\n\t %d secs ago (at %s)\n",
	     afs_cast_time_t(diff), times);
    }

    if (int32p || udebug.amSyncSite) {
	/* now do the subcalls */
	for (i = 0;; i++) {
	    isClone = 0;
	    code = VOTE_XSDebug(tconn, i, &usdebug, &isClone);
	    if (code < 0) {
		if (oldServer) {	/* pre 3.5 server */
		    memset(&usdebug, 0, sizeof(usdebug));
		    code = VOTE_SDebugOld(tconn, i, (ubik_sdebug_old *)&usdebug);
		} else
		    code = VOTE_SDebug(tconn, i, &usdebug);
	    }
	    if (code > 0)
		break;		/* done */
	    if (code < 0) {
		printf("error code %d from VOTE_SDebug\n", code);
		break;
	    }
	    /* otherwise print the structure */
	    printf("\nServer (%s", afs_inet_ntoa(htonl(usdebug.addr)));
	    for (j = 0;
		 ((usdebug.altAddr[j]) && (j < UBIK_MAX_INTERFACE_ADDR - 1));
		 j++)
		printf(" %s", afs_inet_ntoa(htonl(usdebug.altAddr[j])));
	    printf("): (db %d.%d)", usdebug.remoteVersion.epoch,
		   usdebug.remoteVersion.counter);
	    if (isClone)
		printf("    is only a clone!");
	    printf("\n");

	    if (usdebug.lastVoteTime == 0) {
		printf("    last vote never rcvd \n");
	    } else {
		diff = udebug.now - usdebug.lastVoteTime;
		newtime = now - diff;
		times = ctime(&newtime);
		times[24] = 0;
		printf("    last vote rcvd %d secs ago (at %s),\n", 
		       afs_cast_time_t(diff),
		       times);
	    }

	    if (usdebug.lastBeaconSent == 0) {
		printf("    last beacon never sent \n");
	    } else {
		diff = udebug.now - usdebug.lastBeaconSent;
		newtime = now - diff;
		times = ctime(&newtime);
		times[24] = 0;
		printf
		    ("    last beacon sent %d secs ago (at %s), last vote was %s\n",
		     afs_cast_time_t(diff), times, ((usdebug.lastVote) ? "yes" : "no"));
	    }

	    printf("    dbcurrent=%d, up=%d beaconSince=%d\n",
		   usdebug.currentDB, usdebug.up, usdebug.beaconSinceDown);
	}
    }
    return (0);
}

#include "AFS_component_version_number.c"

int
main(int argc, char **argv)
{
    struct cmd_syndesc *ts;

#ifdef	AFS_AIX32_ENV
    /*
     * The following signal action for AIX is necessary so that in case of a 
     * crash (i.e. core is generated) we can include the user's data section 
     * in the core dump. Unfortunately, by default, only a partial core is
     * generated which, in many cases, isn't too useful.
     */
    struct sigaction nsa;

    sigemptyset(&nsa.sa_mask);
    nsa.sa_handler = SIG_DFL;
    nsa.sa_flags = SA_FULLDUMP;
    sigaction(SIGSEGV, &nsa, NULL);
#endif
    ts = cmd_CreateSyntax(NULL, CommandProc, NULL, "probe ubik server");
    cmd_AddParm(ts, "-server", CMD_SINGLE, CMD_REQUIRED, "server machine");
    cmd_AddParm(ts, "-port", CMD_SINGLE, CMD_OPTIONAL, "IP port");
    cmd_AddParm(ts, "-long", CMD_FLAG, CMD_OPTIONAL, "print all info");

    cmd_Dispatch(argc, argv);
    exit(0);
}
